src/ui/speechrecognition/speechrecognizer.js

```diff
@@ -0,0 +1,146 @@
+/* eslint-disable new-cap */
+import { isMicrosoftEdge } from '../../core/utils/useragent';
+import { transformSpeechRecognitionLocaleForEdge } from '../../core/speechrecognition/locales';
+import TranslationFlagger from '../i18n/translationflagger';
+
+/**
+ * Responsible for recognizing speech
+ */
+export default class SpeechRecognizer {
+  constructor (locale) {
+    this._speechRecognition = new webkitSpeechRecognition();
+    this._speechRecognition.interimResults = true;
+
+    this._speechRecognition.lang = isMicrosoftEdge()
+      ? transformSpeechRecognitionLocaleForEdge(locale)
+      : locale;
+
+    /**
+     * The latest result from speech recognition
+     * @type {string}
+     */
+    this._latestResult = '';
+
+    /**
+     * The timeout ID of the silence threshold timeout
+     * @type {number|null}
+     */
+    this._silenceThresholdTimeout = null;
+
+    /**
+     * The amount of silence after the last detected word before triggering a search.
+     * This is primarily used for Safari since Edge and Chrome typically detect the end of
+     * voice search phrase very quickly.
+     */
+    this._silenceThresholdToSearch = 1000;
+
+    /**
+     * Indicates that speech recognition is currently active.
+     * This is used to prevent the speech recognizer from starting while it is already active.
+     * @type {boolean}
+     */
+    this._recognitionActive = false;
+
+    this._init();
+  }
+
+  _init () {
+    this._speechRecognition.addEventListener('start', () => {
+      this._recognitionActive = true;
+    });
+    this._speechRecognition.addEventListener('end', () => {
+      this._recognitionActive = false;
+      this._onCompleteHandler && this._onCompleteHandler(this._latestResult);
+    });
+    this._speechRecognition.addEventListener('result', event => {
+      this._handleResultEvent(event);
+    });
+    this._speechRecognition.addEventListener('error', event => {
+      this._handleErrorEvent(event);
+    });
+  }
+
+  /**
+   * Handles a search recognition result event
+   * @param {SpeechRecognitionEvent} event
+   */
+  _handleResultEvent (event) {
+    /**
+     * Set a timeout to stop the speech recognition if no words are detected for the configured amount
+     * of time. Every time a result comes in, reset that timeout.
+     */
+    clearTimeout(this._silenceThresholdTimeout);
+    this._silenceThresholdTimeout = setTimeout(() => {
+      this.stop();
+    }, this._silenceThresholdToSearch);
+
+    const result = event.results[0][0].transcript;
+    const isFinalResult = event.results[0].isFinal;
+
+    if (result !== this._latestResult) {
+      this._onResultHandler && this._onResultHandler(result);
+      this._latestResult = result;
+    }
+    if (isFinalResult) {
+      this.stop();
+    }
+  }
+
+  /**
+   * Handles a speech recognition error
+   * @param {SpeechRecognitionError} event
+   */
+  _handleErrorEvent (event) {
+    if (event.error === 'service-not-allowed') {
```

Comment:
did you find out whether this error code ever happens for other reasons?

Reply:
Ya, it can happen for a variety of reasons including " because user agent doesn't support it or because of reasons of security, privacy or user preference". Do you think we should include more of that info in our error message? My intent here was to have a general message of 'Speech Recognition is not available.' and also include a message in case the problem was Siri not being enabled

Reply:
ah ok, maybe we could update it slightly with something like `Note: Siri must be ---`? Otherwise to me it implies that the reason voice search isn't working is because of siri, as opposed to a "make sure to check" kind of thing.

Reply:
I was thinking we *could* try sniffing the browser but eh probably not worth the effort/spaghetti

Reply:
I can update the message! To me, sniffing the browser isn't worth it for this